include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,934 @@
+                }
+            }
+        }
+
+        auto newCmd = [&]() -> Cmd*
+        {
+            void* cmdMem = segmentListTail->allocate<Cmd>(args...);
+            if (!cmdMem)
+                return nullptr;
+
+            return new (cmdMem) Cmd(std::forward<Args>(args)...);
         };
 
-        IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx) : IBackendObject(std::move(dev)), m_flags(_flags), m_familyIx(_familyIx) {}
+        auto cmd = newCmd();
+        if (!cmd)
+        {
+            if (!m_commandSegmentPool.appendToList(segmentListHeadItr, segmentListTail))
+                return nullptr;
 
-        core::bitflag<E_CREATE_FLAGS> getCreationFlags() const { return m_flags; }
-        uint32_t getQueueFamilyIndex() const { return m_familyIx; }
+            cmd = newCmd();
+            if (!cmd)
+            {
+                assert(false);
+                return nullptr;
+            }
+        }
 
-        // OpenGL: nullptr, because commandpool doesn't exist in GL (we might expose the internal allocator in the future)
-        // Vulkan: const VkCommandPool*
-        virtual const void* getNativeHandle() const = 0;
+        return cmd;
+    }
 
-    protected:
-        virtual ~IGPUCommandPool() = default;
+    void deleteCommandSegmentList(CCommandSegment::Iterator& segmentListHeadItr, CCommandSegment*& segmentListTail)
+    {
+        // Step #1: Disown the child.
+        if (m_segmentLOLHead && m_segmentLOLTail)
+        {
+            auto current = segmentListHeadItr.segment;
+            auto oneBefore = current->getPrevHead();
+            auto oneAfter = current->getNextHead();
+
+            if (oneAfter)
+                oneAfter->setPrevHead(oneBefore);
+
+            if (oneBefore)
+                oneBefore->setNextHead(oneAfter);
+
+            if (!oneBefore && !oneAfter)
+            {
+                m_segmentLOLHead = nullptr;
+                m_segmentLOLTail = nullptr;
+            }
+        }
+
+        // Step #2: Kill the child.
+        m_commandSegmentPool.deleteList(segmentListHeadItr, segmentListTail);
+    }
+
+    bool reset()
+    {
+        m_resetCount.fetch_add(1);
+        m_commandSegmentPool.clear(m_segmentLOLHead, m_segmentLOLTail);
+        return reset_impl();
+    }
+
+    inline uint32_t getResetCounter() { return m_resetCount.load(); }
+
+protected:
+    IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx)
+        : IBackendObject(std::move(dev)), m_flags(_flags), m_familyIx(_familyIx)
+    {}
+
+    virtual ~IGPUCommandPool() = default;
+
+    virtual bool reset_impl() { return true; };
+
+    core::bitflag<E_CREATE_FLAGS> m_flags;
+    uint32_t m_familyIx;
+
+private:
+    CCommandSegment* m_segmentLOLHead = nullptr;
+    CCommandSegment* m_segmentLOLTail = nullptr;
```

Comment:
the `CCommandSegmentPool` should have been keeping these, and only one of them is needed

Reply:
Resolved.